In [26]:
#instalando e importanto librerías necesarias para el Proyecto
!pip install geopandas
!pip install folium
import numpy as np 
import pandas as pd 
import geopandas as gpd

import matplotlib.pyplot as plt

import folium
from folium import Choropleth, Circle, Marker
from folium.plugins import HeatMap, MarkerCluster


[notice] A new release of pip available: 22.2.2 -> 22.3
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip available: 22.2.2 -> 22.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [27]:
#Importamos el archivo train como data y  test como data_test
data = pd.read_csv(r"C:\Users\Juan\Desktop\PROYECTO INDIVIDUAL\PROYECTO INDIVIDAL 02\Datathon\properties_colombia_train.csv", encoding="utf-8")
data_test = pd.read_csv(r"C:\Users\Juan\Desktop\PROYECTO INDIVIDUAL\PROYECTO INDIVIDAL 02\Datathon\properties_colombia_test.csv")

In [28]:
#Mostrando las 3 primeras filas de data_test
data_test.head(3)

Unnamed: 0                        id    ad_type  start_date    end_date  \
0      852569  pCiVu6rxpBq9dCZfYCXXgA==  Propiedad  2020-09-04  2020-09-12   
1      121379  L0XSQ7mF1TWb1lHRg4n2+A==  Propiedad  2020-10-20  2020-10-20   
2      119504  FbWRxZOMNoUzVOLXgkum1A==  Propiedad  2020-10-20  2020-10-22   

   created_on       lat        lon        l1            l2  ... bathrooms  \
0  2020-09-04  4.722000 -74.075000  Colombia  Cundinamarca  ...       3.0   
1  2020-10-20       NaN        NaN  Colombia     Antioquia  ...       3.0   
2  2020-10-20  6.259253 -75.628381  Colombia     Antioquia  ...       1.0   

  surface_total surface_covered currency  price_period  \
0         660.0           350.0      COP           NaN   
1           NaN             NaN      COP       Mensual   
2           NaN             NaN      COP           NaN   

                                               title  \
0  Casa En Arriendo/venta En Bogota Sotileza Cod....   
1  SE VENDE APARTAMENTO EN EL SECTOR DE ZUÑIGA-EN...   
2           Casa Barata San Javier Solo por este mes   

                                         description  property_type  \
0  EXCLUSIVA CASA EN VENTA Y ARRIENDO, UBICADA EN...           Casa   
1  PR 7555. Apartamento Duplex en unidad cerrada,...    Apartamento   
2  Casa compraventa escrituras en \n proceso,sala...           Casa   

   operation_type                        geometry  
0           Venta           POINT (-74.075 4.722)  
1           Venta                     POINT EMPTY  
2           Venta  POINT (-75.6283806 6.25925267)  

[3 rows x 26 columns]

In [29]:
#conociendo la cantidad de columnas , 
# tipos de datos de cada columna y su cantidad de datos  de data_test 
data_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65850 entries, 0 to 65849
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       65850 non-null  int64  
 1   id               65850 non-null  object 
 2   ad_type          65850 non-null  object 
 3   start_date       65850 non-null  object 
 4   end_date         65850 non-null  object 
 5   created_on       65850 non-null  object 
 6   lat              49433 non-null  float64
 7   lon              49433 non-null  float64
 8   l1               65850 non-null  object 
 9   l2               65850 non-null  object 
 10  l3               62193 non-null  object 
 11  l4               15074 non-null  object 
 12  l5               9052 non-null   object 
 13  l6               2364 non-null   object 
 14  rooms            9110 non-null   float64
 15  bedrooms         13403 non-null  float64
 16  bathrooms        52054 non-null  float64
 17  surface_tota

In [30]:
#mostrando las 3 primeras filas  de data
data.head(3)

Unnamed: 0                        id    ad_type  start_date    end_date  \
0      861131  CScTjaFiP6yCzS8tFX0ZVA==  Propiedad  2020-12-14  9999-12-31   
1      200972  DlCxox9+FuugkxS2kBVfJg==  Propiedad  2020-09-16  2021-05-14   
2      554622  yitN1AigL45gh/K2YSF8rw==  Propiedad  2020-08-19  2020-08-20   

   created_on       lat        lon        l1            l2  ... surface_total  \
0  2020-12-14  6.203000 -75.572000  Colombia     Antioquia  ...           NaN   
1  2020-09-16  4.722748 -74.073115  Colombia  Cundinamarca  ...           NaN   
2  2020-08-19  4.709000 -74.030000  Colombia  Cundinamarca  ...           NaN   

  surface_covered         price currency  price_period  \
0             NaN  9.000000e+09      COP           NaN   
1             NaN  6.250000e+08      COP           NaN   
2             NaN  4.200000e+08      COP           NaN   

                                               title  \
0  Hermosa Mansión En El Poblado. Área Construida...   
1   APARTAMENTO EN VENTA, BOGOTA-CIUDAD JARDIN NORTE   
2  Apartamento En Venta En Bogota Bella Suiza Cod...   

                                         description  property_type  \
0  <b>Hermosa Mansión En El Poblado. Área Constru...           Casa   
1  Venta apartamento en piso 16 con vista exterio...    Apartamento   
2  Al entrar a este bonito apartamento interior e...           Casa   

   operation_type                       geometry  
0           Venta          POINT (-75.572 6.203)  
1           Venta  POINT (-74.0731146 4.7227485)  
2           Venta           POINT (-74.03 4.709)  

[3 rows x 27 columns]

In [31]:
#conociendo la cantidad de columnas , 
# tipos de datos de cada columna y su cantidad de datos de data 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 197549 entries, 0 to 197548
Data columns (total 27 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   Unnamed: 0       197549 non-null  int64  
 1   id               197549 non-null  object 
 2   ad_type          197549 non-null  object 
 3   start_date       197549 non-null  object 
 4   end_date         197549 non-null  object 
 5   created_on       197549 non-null  object 
 6   lat              148051 non-null  float64
 7   lon              148051 non-null  float64
 8   l1               197549 non-null  object 
 9   l2               197549 non-null  object 
 10  l3               186517 non-null  object 
 11  l4               45367 non-null   object 
 12  l5               27409 non-null   object 
 13  l6               6867 non-null    object 
 14  rooms            27537 non-null   float64
 15  bedrooms         40525 non-null   float64
 16  bathrooms        156467 non-null  floa

Analizando promedio de precio de la data original que es lo que  nos piden para poder diferenciar si las propiedas son caras o baratas, el promedio es 6.436051e+08

In [32]:
data["price"].describe()

count    1.974860e+05
mean     6.436051e+08
std      1.944868e+09
min      0.000000e+00
25%      2.000000e+08
50%      3.300000e+08
75%      6.000000e+08
max      3.450000e+11
Name: price, dtype: float64

Seleccionamos las columnas que consideramos más  relevantes al momento de estimar los precios como rooms, bedrooms, bathrooms, surface_covered, price, currency(moneda) y l1(queremos saber si todos perteneces a Colombia)

In [35]:
data1 = data[["rooms","bedrooms","bathrooms","surface_covered","price","currency","l1"]]

Verificamos si es que tanto las monedas y el país corresponden a pesos colombianos (COP) y Colombia, respectivamente.

In [36]:
data1["currency"].value_counts(), data1["l1"].value_counts()

(COP    197474
 USD         8
 Name: currency, dtype: int64,
 Colombia    197549
 Name: l1, dtype: int64)

Vemos que los resultados para data todas las 197,549 propiedades están en Colombia.E n las monedas hay 8 que corresponden a dólares, sólo representa el 0.004% del total,así que lo dejaremos como está.

Ahora veremos la cantida de nulos que hay por cada columna para reemplazarlos por la media, para poder elaborar un buen modelo que nos asegure un buen recall

In [39]:
data1.isna().sum()

rooms              170012
bedrooms           157024
bathrooms           41082
surface_covered    187747
price                  63
currency               67
l1                      0
dtype: int64

Vamos a reemplazar los nulos por valores promedio según el tipo de columna

In [40]:
data1["rooms"] = data1[["rooms"]].fillna(np.mean(data["rooms"]))
data1["bedrooms"] = data1[["bedrooms"]].fillna(np.mean(data["bedrooms"]))
data1["bathrooms"] = data1[["bathrooms"]].fillna(np.mean(data["bathrooms"]))
data1["surface_covered"] = data1[["surface_covered"]].fillna(np.mean(data["surface_covered"]))

C:\Users\Juan\AppData\Local\Temp\ipykernel_15396\3511782626.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data1["rooms"] = data1[["rooms"]].fillna(np.mean(data["rooms"]))
C:\Users\Juan\AppData\Local\Temp\ipykernel_15396\3511782626.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data1["bedrooms"] = data1[["bedrooms"]].fillna(np.mean(data["bedrooms"]))
C:\Users\Juan\AppData\Local\Temp\ipykernel_15396\3511782626.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a Da

Ahora nos queda así del dataframe data1

In [41]:
data1

rooms  bedrooms  bathrooms  surface_covered         price currency  \
0       3.291208  4.000000   4.000000       948.335646  9.000000e+09      COP   
1       3.291208  3.000000   2.000000       948.335646  6.250000e+08      COP   
2       3.291208  3.239062   2.000000       948.335646  4.200000e+08      COP   
3       3.291208  3.239062   1.000000       948.335646  1.310000e+09      COP   
4       3.291208  3.239062   2.000000       948.335646  2.100000e+08      COP   
...          ...       ...        ...              ...           ...      ...   
197544  3.000000  3.000000   2.639279       948.335646  1.290000e+08      COP   
197545  3.291208  3.239062   1.000000        17.000000  1.350000e+08      COP   
197546  3.291208  3.239062   1.000000       948.335646  5.200000e+08      COP   
197547  3.291208  3.239062   2.000000       948.335646  2.900000e+08      COP   
197548  3.291208  3.239062   1.000000       948.335646  3.000000e+07      COP   

              l1  
0       Colombia  
1       Colombia  
2       Colombia  
3       Colombia  
4       Colombia  
...          ...  
197544  Colombia  
197545  Colombia  
197546  Colombia  
197547  Colombia  
197548  Colombia  

[197549 rows x 7 columns]

Sin embargo debemos crear una nueva columna llamada precio_class para diferenciar los precios caros y baratos y luego eliminar las columnas price y currency

In [44]:
data1["precio_class"] = None
data1["precio_class"] = np.where(data1["price"]> 6.436051e+08, "caro", "barato")
data1["precio_class"].value_counts()
data1 = data1[["rooms","bedrooms","bathrooms","surface_covered","precio_class"]]
data1

C:\Users\Juan\AppData\Local\Temp\ipykernel_15396\781604559.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data1["precio_class"] = None
C:\Users\Juan\AppData\Local\Temp\ipykernel_15396\781604559.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data1["precio_class"] = np.where(data1["price"]> 6.436051e+08, "caro", "barato")


rooms  bedrooms  bathrooms  surface_covered precio_class
0       3.291208  4.000000   4.000000       948.335646         caro
1       3.291208  3.000000   2.000000       948.335646       barato
2       3.291208  3.239062   2.000000       948.335646       barato
3       3.291208  3.239062   1.000000       948.335646         caro
4       3.291208  3.239062   2.000000       948.335646       barato
...          ...       ...        ...              ...          ...
197544  3.000000  3.000000   2.639279       948.335646       barato
197545  3.291208  3.239062   1.000000        17.000000       barato
197546  3.291208  3.239062   1.000000       948.335646       barato
197547  3.291208  3.239062   2.000000       948.335646       barato
197548  3.291208  3.239062   1.000000       948.335646       barato

[197549 rows x 5 columns]

El paso final con data 1, vamos a reemplazar barato y caro con 0 y 1 , barato es 0 y caro es 1 

In [45]:
data1 = data1.replace({'barato': 0, 'caro' : 1})
data1

rooms  bedrooms  bathrooms  surface_covered  precio_class
0       3.291208  4.000000   4.000000       948.335646             1
1       3.291208  3.000000   2.000000       948.335646             0
2       3.291208  3.239062   2.000000       948.335646             0
3       3.291208  3.239062   1.000000       948.335646             1
4       3.291208  3.239062   2.000000       948.335646             0
...          ...       ...        ...              ...           ...
197544  3.000000  3.000000   2.639279       948.335646             0
197545  3.291208  3.239062   1.000000        17.000000             0
197546  3.291208  3.239062   1.000000       948.335646             0
197547  3.291208  3.239062   2.000000       948.335646             0
197548  3.291208  3.239062   1.000000       948.335646             0

[197549 rows x 5 columns]

Ahora vamos al data2 que es del test, usaremos las misma columnas 

In [46]:
data2 = data_test[["rooms","bedrooms","bathrooms","surface_covered","currency","l1"]]

Verificamos si es que tanto las monedas y el país corresponden a pesos colombianos (COP) y Colombia, respectivamente

In [47]:
data2["currency"].value_counts(), data2["l1"].value_counts()

(COP    65832
 USD        1
 Name: currency, dtype: int64,
 Colombia    65850
 Name: l1, dtype: int64)

Los resultados nos dicen que los 65,850 registros corresponden a Colombia y con respecto a moneda, de los 65,833 , solamente 1 es de dólar , por lo tanto no influye en el análisis

Haremos un análisis de los datos nulos 

In [49]:
data2.isna().sum()

rooms              56740
bedrooms           52447
bathrooms          13796
surface_covered    62545
currency              17
l1                     0
dtype: int64

Nos damos cuentas que existen muchos datos nulos, por lo tanto aplicaremos el mismo criterio y los reemplazaremos por los promedios de cada columna 

In [50]:
data2["rooms"] = data2[["rooms"]].fillna(np.mean(data2["rooms"]))
data2["bedrooms"] = data2[["bedrooms"]].fillna(np.mean(data2["bedrooms"]))
data2["bathrooms"] = data2[["bathrooms"]].fillna(np.mean(data2["bathrooms"]))
data2["surface_covered"] = data2[["surface_covered"]].fillna(np.mean(data2["surface_covered"]))
data2

C:\Users\Juan\AppData\Local\Temp\ipykernel_15396\2257402602.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data2["rooms"] = data2[["rooms"]].fillna(np.mean(data2["rooms"]))
C:\Users\Juan\AppData\Local\Temp\ipykernel_15396\2257402602.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data2["bedrooms"] = data2[["bedrooms"]].fillna(np.mean(data2["bedrooms"]))
C:\Users\Juan\AppData\Local\Temp\ipykernel_15396\2257402602.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a 

rooms  bedrooms  bathrooms  surface_covered currency        l1
0      3.000000  3.000000   3.000000        350.00000      COP  Colombia
1      3.282327  3.236813   3.000000       5616.33767      COP  Colombia
2      3.282327  3.236813   1.000000       5616.33767      COP  Colombia
3      3.282327  3.236813   2.639624       5616.33767      COP  Colombia
4      3.282327  3.236813   2.000000       5616.33767      COP  Colombia
...         ...       ...        ...              ...      ...       ...
65845  3.000000  3.000000   2.639624       5616.33767      COP  Colombia
65846  3.282327  3.236813   1.000000       5616.33767      COP  Colombia
65847  3.282327  3.236813   2.000000       5616.33767      COP  Colombia
65848  3.282327  3.236813   2.000000       5616.33767      COP  Colombia
65849  3.282327  3.236813   2.000000       5616.33767      COP  Colombia

[65850 rows x 6 columns]

Ahora nos quedamos con las mismas columnas  rooms, bedrooms, bathrooms, surface_covered

In [51]:
data2 = data2[["rooms","bedrooms","surface_covered","bathrooms"]]
data2

rooms  bedrooms  surface_covered  bathrooms
0      3.000000  3.000000        350.00000   3.000000
1      3.282327  3.236813       5616.33767   3.000000
2      3.282327  3.236813       5616.33767   1.000000
3      3.282327  3.236813       5616.33767   2.639624
4      3.282327  3.236813       5616.33767   2.000000
...         ...       ...              ...        ...
65845  3.000000  3.000000       5616.33767   2.639624
65846  3.282327  3.236813       5616.33767   1.000000
65847  3.282327  3.236813       5616.33767   2.000000
65848  3.282327  3.236813       5616.33767   2.000000
65849  3.282327  3.236813       5616.33767   2.000000

[65850 rows x 4 columns]

Ahora pasaremos a la etapa de creación del modelo

In [52]:
from sklearn.ensemble import RandomForestClassifier

n_estimators = cantidad de árboles , se considerará 1000 árboles  ; 
max_features = cantidad de características para cada árbol, se usará sqrt, la raíz cuadrada    ;
bootstrap = muestreo aleatorio de datos , le damos True, aunque viene por defecto      ;
max_samples = procentaje de muestreo aleatorio, usaremos 2/3
oob_score = son los datos que quedaron fuera de los samples

In [53]:
rf = RandomForestClassifier(n_estimators = 1000, 
                            max_features = "sqrt",
                            bootstrap=True, 
                            max_samples= 2/3, 
                            oob_score= True)
data1_f = data1.drop(['precio_class'], axis=1)
rf.fit(data1_f, data1["precio_class"])

RandomForestClassifier(max_samples=0.6666666666666666, n_estimators=1000,
                       oob_score=True)

Entrenamos el modelo  y predecimos en pred

In [54]:
print(rf.predict(data2))
pred = rf.predict(data2)
print(rf.score(data1_f, data1["precio_class"]))
print(rf.oob_score_)

c:\Users\Juan\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names must be in the same order as they were in fit.

  warnings.warn(message, FutureWarning)


[1 1 1 ... 1 1 1]


c:\Users\Juan\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names must be in the same order as they were in fit.

  warnings.warn(message, FutureWarning)


0.8177971035034346
0.8133020162086368


Enviamos los resultados al csv juancortezs07

In [55]:
file = pd.DataFrame(pred, columns = ['pred'])
file.to_csv('juancortezs07.csv', index = False)

FIN